### PR TITLE
Enable bidirectional communication for customEvent

### PR DESCRIPTION
**Pull request by Erli-ms**
### Id of the task, bug, story or other reference.
[User Story 5252317](https://dynamicscrm.visualstudio.com/OneCRM/_workitems/edit/5252317): Support bidirectional Event Messages in LiveChatWidget
### Description
_Include a description of the problem to be solved_
Enable customer to send/receive custom events with bot/agents.

## Solution Proposed
_Detail what is the solution proposed, include links to design document if required or any other document required to support the solution_
Adding Broadcast event channel and event handlers

### Acceptance criteria
_Define what are the conditions to consider the PR has achieved the intended goal_
Events can be received on both users and agent/bot.

## Test cases and evidence
_Include what tests cases were considered, any evidence of testing for future references, to identify any corner cases, etc_

demo: https://ersuoteststorageaccount.blob.core.windows.net/share/demo_video/bidirectional_message_init_demo.mp4

CustomEvent messages would not be displayed in Transcript viewer
image

### Sanity Tests
-   [ ] You have tested all changes in Popout mode
-   [x] You have tested all changes in cross browsers i.e Edge, Chrome, Firefox, Safari and mobile devices(iOS and Android)
-   [x] Your changes are included in the [CHANGELOG](../CHANGE_LOG.md)

## A11y
- [ ] You have run the [Automated Accessibility Check](https://accessibilityinsights.io/docs/en/windows/getstarted/automatedchecks) and have no reported issues

__*Please provide justification if any of the validations has been skipped.*__

### CODE REVIEW

File: chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/storemiddlewares/customEventMiddleware.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+const createCustomEventMiddleware = () => (next: (action: IWebChatAction) => void) => (action: IWebChatAction) => {
+    if (action?.type == WebChatActionType.DIRECT_LINE_INCOMING_ACTIVITY && action.payload?.activity) {
+        console.log("debugging: received incoming activity: ", action);
+        const activity = action.payload.activity;
+        if (activity?.channelData?.metadata?.customEvent 
+            && typeof activity?.channelData?.metadata?.customEvent === "string" && activity?.channelData?.metadata?.customEvent?.toLowerCase() === "true"
+            && activity?.from?.role !== "user"){
+            const customEvent = {
+                eventName: "onCustomEvent",
```
Comment (xTEddie) [inline chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/storemiddlewares/customEventMiddleware.ts:24]: Should the event name be a constant?

File: chat-widget/src/components/livechatwidget/livechatwidgetstateful/LiveChatWidgetStateful.tsx [hunk 2]
```diff
@@ -387,6 +387,52 @@ export const LiveChatWidgetStateful = (props: ILiveChatWidgetProps) => {
             dispatch({ type: LiveChatWidgetActionType.SET_MINIMIZED, payload: msg?.payload?.minimized });
         });
 
+
+
+        BroadcastService.getMessageByEventName("sendCustomEvent").subscribe((event: object) => {
+            let payload = {};
```
Comment (elopezanaya) [inline chat-widget/src/components/livechatwidget/livechatwidgetstateful/LiveChatWidgetStateful.tsx:393]: this logic is not using at all states, I think it will be better to move it to its own file as a hook,

File: chat-widget/src/components/livechatwidget/livechatwidgetstateful/LiveChatWidgetStateful.tsx [hunk 3]
```diff
@@ -387,6 +387,52 @@ export const LiveChatWidgetStateful = (props: ILiveChatWidgetProps) => {
...
+            if ("payload" in event) payload = event.payload as object;
+            console.log("debugging: sendCustomEvent is received", event);
+            if ("customEventName" in payload && payload.customEventName && typeof payload.customEventName === "string" 
+                && "customEventValue" in payload && payload.customEventValue
+            ) {
+                try {
+                    const customEventValueStr = typeof payload.customEventValue === "string" ? payload.customEventValue : JSON.stringify(payload.customEventValue);
```
Comment (elopezanaya) [inline chat-widget/src/components/livechatwidget/livechatwidgetstateful/LiveChatWidgetStateful.tsx:400]: shall we add a util function to validate when an object is string?, is repeated in this logic 3 times

**Review by elopezanaya — CHANGES_REQUESTED** [review]
adjustments suggested

File: CHANGE_LOG.md [hunk 4]
```diff
@@ -38,6 +38,8 @@ All notable changes to this project will be documented in this file.
 - Added public/index.html template with proper styling and configuration guidance for TypeScript sample
 - Added .gitignore file to exclude build artifacts and dependencies from TypeScript sample
 - Added browserslist configuration to TypeScript sample package.json for better browser compatibility
+- Add QueueOverflowHandleMiddleware to display disconnection banner if conversation is ended due to overflow
```
Comment (SamuelBishop) [inline CHANGE_LOG.md:41]: Left over from another PR that must have been in the same branch at least briefly. Please delete this line and let the other PR add it.

Comment (SamuelBishop) [reply CHANGE_LOG.md:41]: Never mind, you are just changing the changelog to make sure the correct h3 has the change message. It is added vs. Changed. Got it. Resolve this one.

**Review by elopezanaya — APPROVED** [review]

**Review by mithunjohny — APPROVED** [review]
